core/transformer/src/ECReferenceTypesCache.ts

```diff
@@ -0,0 +1,168 @@
+    /* eslint-disable quote-props, @typescript-eslint/naming-convention */
+    "Element": ConcreteEntityTypes.Element,
+    "Model": ConcreteEntityTypes.Model,
+    "ElementAspect": ConcreteEntityTypes.ElementAspect,
+    "ElementRefersToElements": ConcreteEntityTypes.Relationship,
+    "ElementDrivesElement": ConcreteEntityTypes.Relationship,
+    // code spec is technically a potential root class but it is sealed and ignored currently
+    // FIXME: because...
+    // "CodeSpec": ConcreteEntityTypes.CodeSpec,
+    /* eslint-enable quote-props, @typescript-eslint/naming-convention */
+  };
+
+  private async getRootBisClass(ecclass: ECClass) {
+    let bisRootForConstraint: ECClass = ecclass;
+    await ecclass.traverseBaseClasses((baseClass) => {
+      // The depth first traversal will descend all the way to the root class before making any lateral traversal
+      // of mixin hierarchies, (or if the constraint is a mixin, it will traverse to the root of the mixin hierarch)
+      // Once we see that we've moved laterally, we can terminate early
+      const isFirstTest = bisRootForConstraint === ecclass;
+      const traversalSwitchedRootPath = baseClass.name !== bisRootForConstraint.baseClass?.name;
+      const stillTraversingRootPath = isFirstTest || !traversalSwitchedRootPath;
+      if (!stillTraversingRootPath)
+        return true; // stop traversal early
+      bisRootForConstraint = baseClass;
+      return false;
+    });
+    // if the root class of the constraint was a mixin, use its AppliesToEntityClass
+    if (bisRootForConstraint instanceof Mixin) {
+      assert(bisRootForConstraint.appliesTo !== undefined, "The referenced AppliesToEntityClass could not be found, how did it pass schema validation?");
+      bisRootForConstraint = await bisRootForConstraint.appliesTo;
+    }
+    return bisRootForConstraint;
+  }
+
+  private async getRootBisClassForConstraint(constraint: RelationshipConstraint) {
+    // constraint classes must share a base so we can get the root from any of them, just use the first
+    const ecclass = await (constraint.constraintClasses?.[0] || constraint.abstractConstraint);
+    assert(ecclass !== undefined, "At least one constraint class or an abstract constraint must have been defined, the constraint is not valid");
+    return this.getRootBisClass(ecclass as ECClass);
+  }
+
+  /** initialize from an imodel with metadata */
+  public async initAllSchemasInIModel(imodel: IModelDb): Promise<void> {
+    const schemaLoader = new SchemaLoader((name: string) => { return imodel.getSchemaProps(name); });
+    await imodel.withPreparedStatement(`
+      SELECT Name FROM ECDbMeta.ECSchemaDef
+      -- schemas defined before biscore are system schemas and no such entities can be transformed so ignore them
+      WHERE ECInstanceId >= (SELECT ECInstanceId FROM ECDbMeta.ECSchemaDef WHERE Name='BisCore')
+      -- ensure schema dependency order
+      ORDER BY ECInstanceId
```

Comment:
This is not guaranteed to be true.  For example if a new system schema is added after an iModel is created it will have an id greater than BisCore but still be referenced by BisCore.  For example the new BisCustomAttributes schema.

That said you can always be sure that a domain schema has a greater id than BisCore

Reply:
See the comment below on where this change comes from. How then do I detect a system schema? I'm going to open up imodelconsole. schemas referenced by biscore?

Reply:
We have in the past hard coded system schema names, but that is not a great solution.  Your query will work 99% of the time since all system schemas that exist at iModel creation time are imported before BisCore.  New 'system' schemas are very rare but the they do happen.

Since not all system schemas are referenced by BisCore (e.g. ECDbMeta) you cannot just look for schemas referenced by BisCore.  

Maybe it's better to look at this in the inverse.  Could you look for schemas that directly or indirectly reference BisCore?  That will always find all domain schemas in an iModel and exclude anything else.  It will exclude schemas that just have custom attributes, units, formats, KOQs, property categories ... but if those schemas are used a domain schema will reference them.

Reply:
replaced with a query to get all descendant schemas of biscore

Reply:
done